matscipy/gamma_surface.py

```diff
@@ -0,0 +1,439 @@
+        self.images = images
+
+    def relax_images(self, calculator=None, ftol=1e-3, optimiser=BFGSLineSearch, constrain_atoms=True,
+                     cell_relax=True, logfile=None, **kwargs):
+        '''
+        Utility function to relax gamma surface images using calculator
+
+        Parameters
+        ----------
+        calculator: ase calculator instance
+            Calculator to give forces for relaxation
+        ftol: float
+            force tolerance for optimisation
+        optimiser: ase optimiser
+            optimiser method to use
+        constrain_atoms: bool
+            Toggle constraint that all atoms only relax in z direction
+            (normal to the gamma surface)
+        cell_relax: bool
+            Allow z component of cell (cell[2, 2]) to relax
+            (normal to the gamma surface)
+        **kwargs: Other keyword args
+            Extra arguments passed to the optimiser.run() method
+        '''
+        if calculator is not None:
+            calc = calculator
+        else:
+            calc = self.calc
+
+        if calc is None:
+            raise RuntimeError("No calculator supplied for relaxation")        
+
+        cell_constraint_mask = np.zeros((3, 3))
+        if cell_relax:
+            cell_constraint_mask[2, 2] = 1  # Allow z component relaxation
+
+        for image in self.images:
+            select_all = np.full_like(image, True, dtype=bool)
+
+            if constrain_atoms:
+                image.set_constraint(FixedLineAtoms(select_all, (0, 0, 1)))
+
+            cell_filter = UnitCellFilter(image, mask=cell_constraint_mask)
+            image.calc = calc
+            opt = optimiser(cell_filter, logfile=logfile)
+            opt.run(ftol, **kwargs)
+
+    def get_surface_energies(self, calculator=None, relax=False, **relax_kwargs):
+        '''
+        Get (self.nx, self.ny) grid of gamma surface energies from self.images
+
+        Parameters
+        ----------
+        calculator : ase calculator
+            Calculator to use for finding surface energies (and optionally in the relaxation)
+        relax : bool
+            Whether to additionally relax the images (through a call to self.relax_images) before finding energies
+        **relax_kwargs : keyword args
+            Extra kwargs to be passed to self.relax_images if relax=True
+
+        Returns 
+        -------
+        Es : np.array
+        (nx, ny) array of energy densities (energy per unit surface area) for the gamma surface, in eV/A**2
+        '''
+
+
+        if calculator is not None:
+            calc = calculator
+        else:
+            calc = self.calc
+
+        if calc is None:
+            raise RuntimeError("No calculator supplied for energy evaluation")
+
+        cell = self.images[0].cell[:, :]
+        self.surface_area = np.linalg.norm(np.cross(cell[0, :], cell[1, :]))
+        self.surface_separation = np.abs(cell[2, 2])
+
+
+        Es = np.zeros((self.nx, self.ny))
```

Comment:
move to __init__

Reply:
Can't go straight to the init, as nx, ny aren't known until image generation. self.Es gets initialised as None